src/FileManager/interactors/downloadBundle.ts

```
@@ -0,0 +1,148 @@
+): Promise<Stream> {
+  const { revision } = params;
+  // is it a revision or not
+  if (revision) {
+    return await downloadWorkingCopy(params);
+  } else {
+    return await downloadReleasedCopy(params);
+  }
+}
+
+async function downloadWorkingCopy(
+  params: DownloadBundleParams,
+): Promise<Stream> {
+  const { requester, learningObjectAuthorUsername, learningObjectId } = params;
+  const learningObject = await getLearningObject(params, true);
+  const hasAccess = authorizeWorkingCopyDownloadRequest(
+    requester,
+    learningObject,
+  );
+  if (!hasAccess) {
+    throw new ResourceError(
+      FileManagerModuleErrorMessages.forbiddenLearningObjectDownload(requester.username),
+      ResourceErrorReason.FORBIDDEN,
+    );
+  }
+  return createBundleStream(learningObject, requester);
+}
+
+async function downloadReleasedCopy(
+  params: DownloadBundleParams,
+): Promise<Stream> {
+  const { requester, learningObjectAuthorUsername, learningObjectId } = params;
+  const learningObject = await getLearningObject(params);
+
+  const fileExists = await Drivers.fileManager().hasAccess({
+    authorUsername: learningObjectAuthorUsername,
+    learningObjectId,
+    learningObjectRevisionId: learningObject.revision,
+    path: 'bundle.zip',
+  });
+  if (fileExists) {
+    return await Drivers.fileManager().streamFile({
+      authorUsername: learningObjectAuthorUsername,
+      learningObjectId,
+      learningObjectRevisionId: learningObject.revision,
+      path: 'bundle.zip',
+    });
+  } else {
+    // if bundle does not exist, create bundle
+    const bundle = await createBundleStream(learningObject, requester);
+    await uploadFile({
+      authorUsername: learningObject.author.username,
+      learningObjectId: learningObject.id,
+      learningObjectRevisionId: learningObject.revision,
+      file: {
+        path: 'bundle.zip',
+        data: bundle,
+      },
+    });
+    // FIXME: catch error thrown and check for NotFound error
```

Should this be deployed without a catch?

---

This type of error handling is not possible with our current download implementation. This is because we open the stream as soon as the route is hit. We can emit the error to the stream. However, the concept of streams does not exist outside of NodeJS. The context of the error is lost during the package transfer. The client will not be able to understand the error that occurred. This should definitely be improved. But, I think that this issue is out of the scope of this PR.